File: echo-go/src/dapp.go
package main

import (
	"encoding/json"
	"io/ioutil"
  "strconv"
  "fmt"
  "log"
  "os"
  "bytes"
  "strings"
  "net/http"
  "dapp/rollups"
)

var (
  infolog  = log.New(os.Stderr, "[ info ]  ", log.Lshortfile)
  errlog   = log.New(os.Stderr, "[ error ] ", log.Lshortfile)
)

func HandleAdvance(data *rollups.AdvanceResponse) error {
  fmt.Printf("Received advance request data %+v\n", data)
	fmt.Println("Adding notice")
  rollupServer := os.Getenv("ROLLUP_HTTP_SERVER_URL")
  if !strings.HasPrefix(rollupServer, "http://") && !strings.HasPrefix(rollupServer, "https://") {
		rollupServer = "http://" + rollupServer
	}

	notice := map[string]string{"payload": data.Payload}
	jsonData, err := json.Marshal(notice)
	if err != nil {
		return err
	}

	response, err := http.Post(rollupServer+"/notice", "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}
	defer response.Body.Close()

	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		return err
	}

	fmt.Printf("Received notice status %d with body %s\n", response.StatusCode, body)
	return nil
}

func HandleInspect(data *rollups.InspectResponse) error {
	fmt.Printf("Received inspect request data %+v\n", data)
	fmt.Println("Adding report")
	rollupServer := os.Getenv("ROLLUP_HTTP_SERVER_URL")
  if !strings.HasPrefix(rollupServer, "http://") && !strings.HasPrefix(rollupServer, "https://") {
		rollupServer = "http://" + rollupServer
	}

	report := map[string]string{"payload": data.Payload}
	jsonData, err := json.Marshal(report)
	if err != nil {
		return err
	}

	response, err := http.Post(rollupServer+"/report", "application/json", bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}
	defer response.Body.Close()

	fmt.Printf("Received report status %d\n", response.StatusCode)
	return nil
}

func Handler(response *rollups.FinishResponse) error {
  var err error

  switch response.Type {
  case "advance_state":
    data := new(rollups.AdvanceResponse)
    if err = json.Unmarshal(response.Data, data); err != nil {
      return fmt.Errorf("Handler: Error unmarshaling advance:", err)
    }
    err = HandleAdvance(data)
  case "inspect_state":
    data := new(rollups.InspectResponse)
    if err = json.Unmarshal(response.Data, data); err != nil {
      return fmt.Errorf("Handler: Error unmarshaling inspect:", err)
    }
    err = HandleInspect(data)
  }
  return err
}

func main() {
  finish := rollups.FinishRequest{"accept"}

  for true {
    infolog.Println("Sending finish")
    res, err := rollups.SendFinish(&finish)
    if err != nil {
      errlog.Panicln("Error: error making http request: ", err)
    }
    infolog.Println("Received finish status ", strconv.Itoa(res.StatusCode))
    
    if (res.StatusCode == 202){
      infolog.Println("No pending rollup request, trying again")
    } else {

      resBody, err := ioutil.ReadAll(res.Body)
      if err != nil {
        errlog.Panicln("Error: could not read response body: ", err)
      }
      
      var response rollups.FinishResponse
      err = json.Unmarshal(resBody, &response)
      if err != nil {
        errlog.Panicln("Error: unmarshaling body:", err)
      }

      finish.Status = "accept"
      err = Handler(&response)
      if err != nil {
        errlog.Println(err)
        finish.Status = "reject"
      }
    }
  }
}

File: echo-cpp/dapp.cpp
#include <stdio.h>
#include <iostream>

#include "3rdparty/cpp-httplib/httplib.h"
#include "3rdparty/picojson/picojson.h"

std::string handle_advance(httplib::Client &cli, picojson::value data) {
    std::cout << "Received advance request data " << data << std::endl;
    std::cout << "Adding notice" << std::endl;
    auto payload = data.get("payload").get<std::string>();
    auto notice = std::string("{\"payload\":\"") + payload + std::string("\"}");
    auto r = cli.Post("/notice", notice, "application/json");
    std::cout << "Received notice status " << r.value().status << " body " << r.value().body << std::endl;
    return "accept";
}

std::string handle_inspect(httplib::Client &cli, picojson::value data) {
    std::cout << "Received inspect request data " << data << std::endl;
    std::cout << "Adding report" << std::endl;
    auto payload = data.get("payload").get<std::string>();
    auto report = std::string("{\"payload\":\"") + payload + std::string("\"}");
    auto r = cli.Post("/report", report, "application/json");
    std::cout << "Received report status " << r.value().status << " body " << r.value().body << std::endl;
    return "accept";
}

int main(int argc, char** argv) {
    std::map<std::string, decltype(&handle_advance)> handlers = {
        {std::string("advance_state"), &handle_advance},
        {std::string("inspect_state"), &handle_inspect},
    };
    httplib::Client cli(getenv("ROLLUP_HTTP_SERVER_URL"));
    cli.set_read_timeout(20, 0);
    std::string status("accept");
    while (true) {
        std::cout << "Sending finish" << std::endl;
        auto finish = std::string("{\"status\":\"") + status + std::string("\"}");
        auto r = cli.Post("/finish", finish, "application/json");
        std::cout << "Received finish status " << r.value().status << std::endl;
        if (r.value().status == 202) {
            std::cout << "No pending rollup request, trying again" << std::endl;
        } else {
            picojson::value rollup_request;
            picojson::parse(rollup_request, r.value().body);

            auto request_type = rollup_request.get("request_type").get<std::string>();
            auto handler = handlers.find(request_type)->second;
            auto data = rollup_request.get("data");
            status = (*handler)(cli, data);
        }
    }
    return 0;
}